x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,370 @@
+	if co.LoadBalanced != nil {
+		cfgp.LoadBalanced = *co.LoadBalanced
+
+		serverOpts = append(
+			serverOpts,
+			WithServerLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
+		connOpts = append(
+			connOpts,
+			WithConnectionLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
 	}
 
-	var currentBlock *pem.Block
-	var certBlock, certDecodedBlock, keyBlock []byte
+	serverOpts = append(
+		serverOpts,
+		WithClock(func(*session.ClusterClock) *session.ClusterClock { return clusterClock }),
+		WithConnectionOptions(func(...ConnectionOption) []ConnectionOption { return connOpts }))
 
-	remaining := data
-	start := 0
-	for {
-		currentBlock, remaining = pem.Decode(remaining)
-		if currentBlock == nil {
-			break
-		}
+	cfgp.ServerOpts = serverOpts
 
-		if currentBlock.Type == "CERTIFICATE" {
-			certBlock = data[start : len(data)-len(remaining)]
-			certDecodedBlock = currentBlock.Bytes
-			start += len(certBlock)
-		} else if strings.HasSuffix(currentBlock.Type, "PRIVATE KEY") {
-			if keyPasswd != "" && x509.IsEncryptedPEMBlock(currentBlock) {
-				var encoded bytes.Buffer
-				buf, err := x509.DecryptPEMBlock(currentBlock, []byte(keyPasswd))
-				if err != nil {
-					return "", err
-				}
-
-				pem.Encode(&encoded, &pem.Block{Type: currentBlock.Type, Bytes: buf})
-				keyBlock = encoded.Bytes()
-				start = len(data) - len(remaining)
-			} else {
-				keyBlock = data[start : len(data)-len(remaining)]
-				start += len(keyBlock)
-			}
+	// Deployment
+	if co.Deployment != nil {
+		// topology options: WithSeedlist, WithURI, WithSRVServiceName, WithSRVMaxHosts, and WithServerOptions
+		// server options: WithClock and WithConnectionOptions + default maxPoolSize
+		if len(serverOpts) > 2+defaultOptions || hasOptionalData(*co) {
+			return nil, errors.New("cannot specify topology or server options with a deployment")
 		}
 	}
-	if len(certBlock) == 0 {
-		return "", fmt.Errorf("failed to find CERTIFICATE")
-	}
-	if len(keyBlock) == 0 {
-		return "", fmt.Errorf("failed to find PRIVATE KEY")
-	}
 
-	cert, err := tls.X509KeyPair(certBlock, keyBlock)
-	if err != nil {
-		return "", err
-	}
-
-	cfg.Certificates = append(cfg.Certificates, cert)
+	return cfgp, nil
+}
 
-	// The documentation for the tls.X509KeyPair indicates that the Leaf certificate is not
-	// retained.
-	crt, err := x509.ParseCertificate(certDecodedBlock)
-	if err != nil {
-		return "", err
-	}
+// NewConfig will translate client options into a topology config for building deployments.
+func NewConfig(co *options.ClientOptions) (*Config, error) {
+	return newConfig(co, nil)
+}
 
-	return crt.Subject.String(), nil
+// NewConfigWithClient will translate client options into a topology config for building deployments. It will also use
+// a client implementation for accessing and setting unexported data as well as implementing non-configuration-specific
+// data.
+func NewConfigWithClient(co *options.ClientOptions, client client) (*Config, error) {
```

Comment:
This function will likely become unnecessary if we remove the `client` interface and just allow callers to pass in a nil-able `session.ClusterClock`.